### DISCUSSION TITLE
Rye vs Pipx?

### DISCUSSION THREAD

**Discussion by AlJohri** [pypa/pipx#1079] [General]
One of the my biggest issues with `pipx` has been that it doesn't own its own python binary (https://github.com/pypa/pipx/issues/244). Today I found out that [rye](https://github.com/mitsuhiko/rye) solves this problem and can serve as a potential replacement for pipx. Check the [Global Tools](https://github.com/mitsuhiko/rye#global-tools) section in the README:

>If you want tools to be installed into isolated virtualenvs (like pipsi and pipx), you can use rye too (requires ~/.rye/shims to be on the path):
> ```
>$ rye install pycowsay
>$ pycowsay Wow
>
>  ---
>< Wow >
>  ---
>   \   ^__^
>    \  (oo)\_______
>       (__)\       )\/\
>           ||----w |
>           ||     ||
> ```
> To uninstall run rye uninstall pycowsay again.

It apparently uses this python:

>No System Python: I can't deal with any more linux distribution weird Python installations or whatever mess there is on macOS. I used to build my own Pythons that are the same everywhere, now I use [indygreg's Python builds](https://github.com/indygreg/python-build-standalone). Rye will automatically download and manage Python builds from there. No compiling, no divergence.

Example showing how `rye` manages its own python:
```shell
❯ rye toolchain fetch 3.10
Downloading cpython@3.10.13
Checking checksum
success: Downloaded cpython@3.10.13

❯ rye toolchain list
cpython@3.12.0 (/home/aljohri/.rye/py/cpython@3.12.0/install/bin/python3)
cpython@3.11.6 (/home/aljohri/.rye/py/cpython@3.11.6/install/bin/python3)
cpython@3.10.13 (/home/aljohri/.rye/py/cpython@3.10.13/install/bin/python3)
...
```

Example installing global tools with rye:

```shell
❯ rye install csvkit
...
Installed scripts:
  - csvclean
  - csvcut
  - csvformat
  - csvgrep
  - csvjoin
  - csvjson
  - csvlook
  - csvpy
  - csvsort
  - csvsql
  - csvstack
  - csvstat
  - in2csv
  - sql2csv

note: additional scripts were encountered in non-installed dependencies.

❯ which csvcut
~/.rye/shims/csvcut

❯ csvcut --version
csvcut 1.2.0
```

What do other folks think of this approach? Anyone else using `rye` over `pipx`? Can `pipx` also stop relying on the system python?

**Comment by dukecat0** [comment]
> Can pipx also stop relying on the system python?

I had a thought of using [`pyapp`](https://github.com/ofek/pyapp) to build binary distributions which have self-contained Python, but looks like folks aren't happy with it. https://github.com/pypa/pipx/issues/278#issuecomment-1618911330

**Reply by uranusjr** [reply]
My feeling is as long as `PIPX_PYTHON` is available, using python-build-standalone by default should be a good idea. This would resolve the pipx installation problem as well since we no longer need to worry about accomodating users installing pipx against all different kinds of Python distributions.

**Comment by alextremblay** [comment]
i'm ok with pipx itself being installed onto a system python, but i would love it if pipx could manage its own python versions the way rye does.

```
$ ls ~/.local/pipx/
logs py shared  venvs

$ ls ~/.local/pipx/py
cpython@3.10.13  cpython@3.11.6  cpython@3.12.0
```

I think it would be amazing if you could do `pipx run --python=3.11 cowsay -t hello` and have pipx download cpython@3.11.6 from [indygreg's Python builds](https://github.com/indygreg/python-build-standalone) (if it hasn't already) and then use that python installation's venv to create a new pipx venv for `cowsay`

**Reply by alextremblay** [reply]
in fact, this may already be possible... there's a lovely little library called [yen](https://github.com/tusharsadhwani/yen) that can manage python installations in a specific folder, and it uses indygreg's python builds 

it's MIT-licensed, so perhaps pipx could vendor all or part of yen's implementation...

If any of the maintainers are interested in this, let me know, I'll start working on a PR for it :)
